src/mongo/db/storage/csv_file.cpp

```diff
@@ -441,28 +458,88 @@ boost::optional<BSONObj> CsvFileInput::readBsonObj() {
+ * +--------------+       cur_char = "            +----------------+
+ * |     State:   | ----------------------------> |      State:    | other characters
+ * |  quote_open  |                               | checkForDouble | ---> RFC Non compliant
+ * |              | <---------------------------- |     -Quote     |
+ * +--------------+       cur_char = "            +----------------+
+ *    |   ⬆︎          Double double quote (""),
+ *    |   |           Escaped double quote
+ *    |___|
+ *   Any character, except double quote "
+ */
+
 std::vector<std::string_view> CsvFileInput::parseLine(const std::string_view& line) {
-    int left = 0;
-    int right = 0;
-    int len = line.length();
-    int i = 0;
+    auto stateMachine = parsingState::beginField;
+
+    const size_t len = line.length();
+    size_t i = 0;
+    size_t left = 0;
     std::vector<std::string_view> fields;
 
     while (i <= len) {
-        if (i == len || line.at(i) == ',') {
-            fields.emplace_back(line.substr(left, right - left));
-            left = 0;
-            right = 0;
-        } else if (line.at(i) != ' ' && line.at(i) != '\t') {
-            if (right == 0) {
+        switch (stateMachine) {
+            case parsingState::beginField:
                 left = i;
-            }
-            right = i + 1;
+                if (i == len || line[i] == ',') {  // Empty field.
+                    fields.emplace_back(line.substr(left, 0));
+                    left++;
+                    break;
+                }
+                stateMachine =
+                    line[i] == '\"' ? parsingState::quoteOpen : parsingState::quoteClosed;
+                break;
+            case parsingState::quoteOpen:
+                if (i == len) {  // Field that does not comply with RFC 4180
+                    _ioStats->incNonCompliantWithRFC();
+                    return {};
+                }
+                if (line[i] == '\"') {
+                    stateMachine = parsingState::checkForDoubleQuote;
+                }
+                break;
+            case parsingState::quoteClosed:
+                if (i == len || line[i] == ',') {  // End of Field.
+                    fields.emplace_back(line.substr(left, i - left));
+                    stateMachine = parsingState::beginField;
+                } else if (line[i] == '\"') {  // Field that does not comply with RFC 4180
+                    _ioStats->incNonCompliantWithRFC();
+                    return {};
+                }
+                break;
+            case parsingState::checkForDoubleQuote:
+                if (i == len || line[i] == ',') {  // End of field.
+                    fields.emplace_back(line.substr(left + 1, i - 2 - left));
+                    stateMachine = parsingState::beginField;
+                } else if (line[i] == '\"') {  // Escaping double quote.
+                    stateMachine = parsingState::quoteOpen;
+                } else {  // Field that does not comply with RFC 4180
```

Comment:
[nit] The comment is repeating _ioStats->incNonCompliantWithRFC() and let's remove it. The less code, the better if the same thing can be achieved.

```suggestion
                } else {  // Field that does not comply with RFC 4180
```

Reply:
Done